### PR TITLE
Dynamic store callbacks

### PR DESCRIPTION
**Pull request by MxtOUT**
- Simplifies the cross-context messaging by creating simple-to-use interfaces for both sides
- Adds a listener that will get triggered once DynamicStore has finished loading, thus making it easy to determine the time when to start highlighting
- Unrelated addition of the giant curator capsule on the front page to highlighting

### CODE REVIEW

File: js/content/common.js [hunk 1]
```diff
@@ -196,16 +196,45 @@ let ExtensionLayer = (function() {
...
+        script.textContent = '(' + fun + ")();";
         document.documentElement.appendChild(script);
         script.parentNode.removeChild(script);
     };
 
     return self;
 })();
 
+class Messenger {
+    static sendMessage(msgID, info) {
```
Comment (KarlCastle) [inline js/content/common.js]: There is a "sendMessage" API (in chrome.runtime) that is a separate thing from the "postMessage/onmessage" API (on Window/Worker), should we keep this as `postMessage()`? (Or just `post()`, and then `addListener()`?)

File: js/content/common.js [hunk 2]
```diff
@@ -196,16 +196,45 @@ let ExtensionLayer = (function() {
...
+        window.postMessage({
+            type: "es_" + msgID,
+            information: info
+        }, '*');
```
Comment (KarlCastle) [inline js/content/common.js]: It's good practice to *always* specify the `targetOrigin` when you know it, and we know it: `window.location.origin`

File: js/content/common.js [hunk 3]
```diff
@@ -196,16 +196,45 @@ let ExtensionLayer = (function() {
...
+    }
+
+    static addMessageListener(msgID, fn, autoRemove) {
```
Comment (KarlCastle) [inline js/content/common.js]: The option on `addEventListener()` to only fire once is called `once`, should we be using the same name here?

File: js/content/common.js [hunk 4]
```diff
@@ -196,16 +196,45 @@ let ExtensionLayer = (function() {
...
+        let callback = function(e) {
+            if (e.source !== window) { return; }
+            if (!e.data.type) { return; }
```
Comment (KarlCastle) [inline js/content/common.js]: `e.data.type` has the potential to throw if someone does something silly like: `window.postMessage(null)`. Maybe check `e.data` is meaningful?

File: js/content/common.js [hunk 5]
```diff
@@ -196,16 +196,45 @@ let ExtensionLayer = (function() {
...
+            }
+        }
+        window.addEventListener("message", callback);
+    }
+}
+
+// Inject the Messenger class into the DOM, providing the same interface for the page context side
+(function() {
+    let script = document.createElement("script");
+    script.textContent = Messenger;
```
Comment (KarlCastle) [inline js/content/common.js]: This is an implicit `.toString()`, can we make it explicit?

File: js/content/common.js [hunk 6]
```diff
@@ -196,16 +196,45 @@ let ExtensionLayer = (function() {
...
+            if (e.data.type === "es_" + msgID) {
+                fn(e.data.information);
+                if (autoRemove) {
+                    window.removeEventListener("message", callback);
+                }
+            }
+        }
+        window.addEventListener("message", callback);
+    }
+}
```
Comment (KarlCastle) [inline js/content/common.js]: Hmm. Do we do this:
```
Messenger.sendMessage("message", data);
Messenger.addEventListener("message", callback, true);
```
often enough that it makes sense for a response callback parameter on `sendMessage()`? ie:
```
static sendMessage(msgID, info, fn) {
  if (fn) {
     Messenger.addMessageListener(msgID, fn, true);
  }
  ...
}
```

File: js/content/store.js [hunk 7]
```diff
@@ -2752,23 +2752,21 @@ let WishlistPageClass = (function(){
...
-                    location.reload();
-                });
+            let promises = [];
+            for (let i=0; i<wishlistData.length; i++) {
+                let appid = wishlistData[i].appid;
+                promises.push(removeApp(appid));
             }
-        }, false);
+
+            Promise.all(promises).finally(() => {
```
Comment (KarlCastle) [inline js/content/store.js]:
```
Promise.all(wishlistData.map(app => removeApp(app.appid))).finally(() => {
```

File: js/content/store.js [hunk 8]
```diff
@@ -2852,22 +2850,12 @@ let WishlistPageClass = (function(){
         ExtensionLayer.runInPageContext(() =>
             $J(document).ajaxSuccess(function( event, xhr, settings ) {
                 if (settings.url.endsWith("/remove/")) {
-                    window.postMessage({
-                        type: "es_remove_wl_entry",
-                        removed_wl_entry: settings.data.match(/(?!appid=)\d+/)[0]
-                    }, "*");
+                    Messenger.sendMessage("removeWlEntry", settings.data.match(/(?!appid=)\d+/)[0]);
```
Comment (KarlCastle) [inline js/content/store.js]: The message text can be anything, and can include spaces. I'm a fan of "As explicit and readable as possible", so "remove-wishlist-entry" or "remove wishlist entry" or anything without an abbreviation all seem better than "removeWlEntry".

File: js/content/common.js (hunk 2 above)
Comment (MxtOUT) [reply js/content/common.js]: I thought about this too, but doesn't the extension context have a different origin?

File: js/content/common.js (hunk 1 above)
Comment (MxtOUT) [reply js/content/common.js]: `postMessage()` sounds good

File: js/content/common.js (hunk 4 above)
Comment (MxtOUT) [reply js/content/common.js]: Better safe than sorry I guess

File: js/content/common.js (hunk 3 above)
Comment (MxtOUT) [reply js/content/common.js]: Sure

File: js/content/common.js (hunk 6 above)
Comment (MxtOUT) [reply js/content/common.js]: I think we're only doing this once with the user notes, where the page context sends a message before the modal is trying to get dismissed.
Then the extension saves the note and replies with an "OK" signal, so the modal can be closed now.
So I don't think it's worth it

File: js/content/common.js (hunk 5 above)
Comment (MxtOUT) [reply js/content/common.js]: Yes

File: js/content/store.js (hunk 8 above)
Comment (MxtOUT) [reply js/content/store.js]: I'd go for "removeWishlistEntry" here just to keep in style with the other messages

**Pull request merged**